README.md
# Python Docstring Formatter Extension

Simple extension that implements
[`docformatter`](https://pypi.org/project/docformatter/) as a native VSCode
formatter. The formatter uses
[PEP 257](https://www.python.org/dev/peps/pep-0257/) as a guide.

[![Build Status](https://dev.azure.com/iansan5653/vscode-format-python-docstrings/_apis/build/status/iansan5653.vscode-format-python-docstrings?branchName=master)](https://dev.azure.com/iansan5653/vscode-format-python-docstrings/_build/latest?definitionId=1&branchName=master)
[![Visual Studio Marketplace Installs](https://img.shields.io/visual-studio-marketplace/i/iansan5653.format-python-docstrings)](https://marketplace.visualstudio.com/items?itemName=iansan5653.format-python-docstrings)
[![Greenkeeper badge](https://badges.greenkeeper.io/iansan5653/vscode-format-python-docstrings.svg)](https://greenkeeper.io/)

## Features

See https://pypi.org/project/docformatter/ for the full feature list. This
extension uses that project as a backend with few modifications.

All relevant flags are implemented as [extension settings](#extension-settings).

## How to Use

If you already have a Python formatting extension installed, you will need to
pick this one when you want to format a file. In a Python file in VSCode,
<kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>p</kbd> to open the command pallette,
run the command _Format Document With..._ and select
_Python Docstring Formatter_.

To set the default formatter that is run when you run the _Format Document_,
on Python files, add the following to your **settings.json** (`...` being your other settings):

```json
{
    ...,
    "[python]": {
        "editor.defaultFormatter": "iansan5653.format-python-docstrings"
    }
}
```

_Note_: The _Python Docstring Formatter_ option will only appear for files
that have been saved on disk (ie, not **Untitled-#** tabs).

_Note_: The extension will use the Python you have set in `python.pythonPath`.
It does not support Python 2, so this must be set to some version of Python 3.
If the setting is empty, the extension will use a global Python installation.

## Requirements

This extension requires `pip` and/or the `docformatter` module to be installed.
Upon activating, the extension will automatically give the option to run
`pip install --upgrade docformatter` if necessary. If you prefer to install the
formatter program in another way, you may do so. Just make sure that the
`docformatter` command is recognized by your terminal (ie, added to your PATH).

## Extension Settings

All relevant flags from the original `docformatter` program are implemented as
settings:

- `docstringFormatter.wrapSummariesLength` (`number`, default `79`): Wrap long
  summary lines at this length; set to `0` to disable wrapping.
- `docstringFormatter.wrapDescriptionsLength` (`number`, default `72`): Wrap
  descriptions at this length; set to `0` to disable wrapping.
- `docstringFormatter.preSummaryNewline` (`boolean`, default `false`): Add a
  newline before the summary of a multi-line docstring.
- `docstringFormatter.makeSummaryMultiline` (`boolean`, default `false`): Add a
  newline before and after the summary of a one-line docstring.
- `docstringFormatter.forceWrap` (`boolean`, default `false`): Force
  descriptions to be wrapped even if it may result in a mess.

## Release Notes

See [CHANGELOG.md](CHANGELOG.md) for release notes.

## Contributing

Before contributing, you will need to install
[Node.js (with NPM)](https://nodejs.org/en/) (and Visual Studio Code, of course).
Clone the repository and install the required dependencies by running
`npm install` in a terminal in the new directory. All extension logic is in
`src/extension.ts`. Whenever modifying extension behavior, be sure to update and
add unit tests in `src/test/extension.test.ts`. To run/test the extension,
simply use the debugger in VSCode after opening the extension folder.


CHANGELOG.md
# Change Log

All notable changes to the "format-python-docstrings" extension will be
documented in this file.

<!-- ## X.X.X [Unreleased] -->

## 1.2.1

### Fixed

- Support workspace variables in `python.pythonPath` setting (#36)
- Support local paths in `python.pythonPath` setting (#36)
- Check to ensure that the Python version being used is `3.x.x` (#40)
- Check for `python3` command as a possible Python command (#40)

## 1.2.0

### Added

- Respect the `python.pythonPath` setting
- Provide more helpful error messages

## 1.1.0

### Fixed

- Fix failure when attempting to format files which contain spaces in their path
- Update dependencies

## 1.0.2

### Added

- Added button to submit bug report in error messages

### Changed

- Use `compress-tag` module instead of custom method
- Improve marketplace listing style
- Format code and ensure code quality
- Update dependencies

## 1.0.1

### Changed

- Update dependencies
- Implement new `vscode-test` library
- Improve code quality and format

## 1.0.0

### Changed

- Version bumped to 1.0.0 out of preview, absent any bug reports or feature
  requests. No additional features are planned at the time.

## 0.6.1

### Changed

- Update dependencies (to resolve vulnerabilities)

## 0.6.0

### Changed

- Dispose of formatter registration upon deactivation

## 0.5.2

### Added

- Added new changelog

### Changed

- Updated readme
- Fixed text color in marketplace
- Made settings descriptions easier to understand

## 0.5.1

### Added

- Added an icon
- Marketplace metadata to help discovery

## 0.5.0

Preview release. May contain bugs; if you encounter any issues or have feature
requests, please submit an issue on this repository!


src/test/index.ts
import * as path from "path";
import * as Mocha from "mocha";
import * as glob from "glob";

export function run(
  _: unknown,
  cb: (error: null | Error, failures?: number) => void
): void {
  // Create the mocha test
  const mocha = new Mocha({
    ui: "bdd",
    reporter: "mocha-junit-reporter",
    reporterOptions: {
      mochaFile: path.resolve(__dirname, "./results.xml"),
      includePending: true
    }
  });

  const testsRoot = path.resolve(__dirname, "..");
  glob("**/**.test.js", {cwd: testsRoot}, (err, files): void => {
    if (err) {
      return cb(err);
    }

    // Add files to the test suite
    files.forEach((f): Mocha => mocha.addFile(path.resolve(testsRoot, f)));

    try {
      // Run the mocha test
      mocha.run((failures): void => {
        console.log("##vso[task.setvariable variable=testsRan]true");
        console.log(`${failures} test(s) failed.`);
        cb(null, failures);
      });
    } catch (err) {
      console.log("##vso[task.setvariable variable=testsRan]false");
      console.log("Failed to run tests");
      cb(err);
    }
  });
}


src/test/example.py
"""Example file for testing."""


def function_with_good_docstring(param: str) -> bool:
    """This is a well-formatted docstring.

    It has a longer description here as well.

    # Arguments
        param (str): A parameter that is passed to the function.

    # Returns
        bool: A boolean value.
    """
    return bool(param)


def function_with_bad_docstring(param: str) -> bool:
    '''
This is a badly-formatted docstring.
    It has a longer description here as well. But it's way too long to fit the line that it is in. It just keeps going on and on. And the author didn't put any line breaks in.

    It should not use single quotes or have trailing whitespace.

    # Arguments
        param (str): A parameter that is passed to the function.

    # Returns       
        bool: A boolean value. '''
    return bool(param)


src/test/run-tests.ts
import * as path from "path";
import {runTests} from "vscode-test";

async function main(): Promise<void> {
  try {
    const extensionDevelopmentPath = path.resolve(__dirname, "../../");
    const extensionTestsPath = path.resolve(__dirname, "./index");

    await runTests({
      version: "stable",
      extensionDevelopmentPath,
      extensionTestsPath,
      launchArgs: ["--disable-extensions"]
    });
  } catch (err) {
    console.error("Failed to run tests");
    process.exit(1);
  }
}

main();


src/extension.ts
import * as vsc from "vscode";
import * as util from "util";
import * as cp from "child_process";
import * as diff from "diff";
import {c} from "compress-tag";
import * as path from "path";

export const registrations: vsc.Disposable[] = [];

/**
 * Replaces the standard VSCode variables `${workspaceFolder}` and
 * `${workspaceFolderBasename}` with their correct values. Supports the
 * `${workspaceFolder:folderName}` as defined in the spec for multi-root
 * workspaces.
 * @param text The text to replace values in.
 * @returns The input string with the variables replaced.
 */
export function replaceWorkspaceVariables(text: string): string {
  const workspaceFolders = vsc.workspace.workspaceFolders;
  const result = text.replace(
    /\${workspaceFolder(Basename)?:?([^}]*)?}/g,
    (_, baseOnly?: string, dirName?: string): string => {
      const targetFolder = (dirName
        ? workspaceFolders?.find((folder): boolean => folder.name === dirName)
        : workspaceFolders?.[0]
      )?.uri.fsPath;
      return (
        (baseOnly ? path.basename(targetFolder ?? "") : targetFolder) ?? ""
      );
    }
  );
  return result;
}

/**
 * Promisified `exec` function, but with `cwd` bound to a workspace directory.
 * @param command The command to run.
 * @param opts Any other options to pass to `cp.exec`.
 * @param workspaceFolder The name of the workspace folder to bind the `cwd`
 * setting to. If not provided, will use the first folder in the workspace.
 * @returns Promise that resolves with the output of running the command.
 */
export function promiseExec(
  command: string,
  opts?: cp.ExecOptions,
  workspaceFolder?: string
): Promise<{stdout: string; stderr: string}> {
  const folderName = workspaceFolder ? `:${workspaceFolder}` : "";
  const workspaceDirectory = // undefined if replaceWorkspaceVariables yields ""
    replaceWorkspaceVariables(`\${workspaceFolder${folderName}}`) || undefined;
  return util.promisify(cp.exec)(command, {cwd: workspaceDirectory, ...opts});
}

/** Returns the set Python path from settings without replacing variables. */
export function getPythonPathSetting(): string | undefined {
  return vsc.workspace.getConfiguration("python").get<string>("pythonPath");
}

/**
 * Returns `true` if the command points to a valid executable for Python 3.
 * @param command The command to try, ie `python`.
 */
export async function pointsToPython3(command: string): Promise<boolean> {
  try {
    const result = await promiseExec(`${command} --version`);
    return !result.stderr && result.stdout.includes("Python 3");
  } catch {
    return false;
  }
}

/**
 * Get the path to the most locally set Python. If Python cannot be found, this
 * will reject. Otherwise, will resolve with the python command.
 * @param setPath The set Python path. Defaults to user setting. This allows for
 * testing.
 * @returns A command which can be called to invoke Python in the terminal.
 */
export async function getPython(
  setPath: string | undefined = getPythonPathSetting()
): Promise<string> {
  let errorMessage: string;

  if (setPath !== undefined) {
    const cookedPath = `"${replaceWorkspaceVariables(setPath)}"`;
    if (await pointsToPython3(cookedPath)) return cookedPath;
    errorMessage = c`
      The Python path set in the "python.pythonPath" setting is invalid or
      points to Python 2. Check the value or clear the setting to use the
      global Python installation. The extension is not compatible with Python
      2.
    `;
  } else {
    for (const command of ["python", "python3", "py"]) {
      if (await pointsToPython3(command)) return command;
    }
    errorMessage = c`
      Python 3 is either not installed or not properly configured. Check
      that the Python location is set in VSCode or provided in the system
      environment variables.
    `;
  }

  vsc.window.showErrorMessage(errorMessage);
  throw new Error(errorMessage);
}

/**
 * Build a text string that can be run as the Docformatter command with flags.
 *
 * Reads the current settings and implements them or falls back to defaults.
 * @param path Path to the file to be formatted.
 * @param pythonPromise Promise that resolves to the path to call Python with.
 * Allows for unit testing.
 * @returns Runnable terminal command that will format the specified file.
 */
export async function buildFormatCommand(
  path: string,
  pythonPromise: Promise<string> = getPython()
): Promise<string> {
  const python = await pythonPromise;
  const settings = vsc.workspace.getConfiguration("docstringFormatter");
  // Abbreviated to keep template string short
  const wsl = settings.get<number>("wrapSummariesLength") || 79;
  const wdl = settings.get<number>("wrapDescriptionsLength") || 72;
  const psn = settings.get<boolean>("preSummaryNewline") || false;
  const msn = settings.get<boolean>("makeSummaryMultiline") || false;
  const fw = settings.get<boolean>("forceWrap") || false;
  return c`
    ${python} -m
    docformatter "${path}" --wrap-summaries ${wsl} --wrap-descriptions ${wdl}
    ${psn ? "--blank" : ""}
    ${msn ? "--make-summary-multi-line" : ""}
    ${fw ? "--force-wrap" : ""}
  `
    .trim()
    .replace(/\s+/, " "); // Remove extra whitespace (helps with tests)
}

/**
 * Installs Docformatter on the current system, assuming pip is installed.
 * @returns Empty promise that resolves upon succesful installation.
 */
export async function installDocformatter(): Promise<void> {
  const python = await getPython();
  try {
    await promiseExec(`${python} -m pip install --upgrade docformatter`);
  } catch (err) {
    vsc.window.showErrorMessage(c`
      Could not install docformatter automatically. Make sure that pip
      is installed correctly and try manually installing with 'pip
      install --upgrade docformatter. \n\n Full error: ${err}'.
    `);
    throw err;
  }
}

/**
 * Handle an error raised by `promiseExec`, which passes an exception object.
 *
 * If the error signifies that Docformatter is not installed, will attempt
 * to automatically install the tool if the user desires.
 *
 * @param err The error raised by `promiseExec`, which would have been run to
 * execute the format command.
 */
export async function alertFormattingError(
  err: FormatException
): Promise<void> {
  if (err.message.includes("No module named docformatter")) {
    const installButton = "Install Module";
    const response = await vsc.window.showErrorMessage(
      c`The Python module 'docformatter' must be installed to format
          docstrings.`,
      installButton
    );
    if (response === installButton) {
      installDocformatter();
    }
  } else {
    const bugReportButton = "Submit Bug Report";
    const response = await vsc.window.showErrorMessage(
      c`Unknown Error: Could not format docstrings. Full error:\n\n
        ${err.message}`,
      bugReportButton
    );
    if (response === bugReportButton) {
      vsc.commands.executeCommand(
        "vscode.open",
        vsc.Uri.parse(
          "https://github.com/iansan5653/vscode-format-python-docstrings/issues/new"
        )
      );
    }
  }
}

/**
 * Format a file using Docformatter and return the edit hunks without
 * modifying the file.
 * @param path Full path to a file to format. Defaults to the currently open
 * file, if there is one. If not given and no currently open file is present,
 * will raise an error.
 * @returns A promise that resolves to the edit hunks, which can then be
 * converted to edits and applied to the file. If the promise rejects, will
 * automatically show an error message to the user.
 */
export async function formatFile(
  path: string | undefined = vsc.window.activeTextEditor?.document.fileName
): Promise<diff.Hunk[]> {
  if (!path) {
    throw new Error(
      "Could not format docstrings because no file path was given."
    );
  }

  try {
    const command = await buildFormatCommand(path);
    const result = await promiseExec(command);
    const parsed = diff.parsePatch(result.stdout);
    return parsed[0].hunks;
  } catch (err) {
    alertFormattingError(err);
    throw err;
  }
}

/**
 * Convert any number of hunks to a matching array of native VSCode edits.
 * @param hunks Array of hunks to convert to edits.
 * @returns Array of VSCode text edits, which map directly to the input hunks.
 */
export function hunksToEdits(hunks: diff.Hunk[]): vsc.TextEdit[] {
  return hunks.map((hunk) => {
    const startPos = new vsc.Position(hunk.newStart - 1, 0);
    const endPos = new vsc.Position(
      hunk.newStart - 1 + hunk.oldLines - 1,
      hunk.lines[hunk.lines.length - 1].length - 1
    );
    const editRange = new vsc.Range(startPos, endPos);

    const newTextLines = hunk.lines
      .filter(
        (line): boolean => line.charAt(0) === " " || line.charAt(0) === "+"
      )
      .map((line): string => line.substr(1));
    const lineEndChar: string = hunk.linedelimiters[0];
    const newText = newTextLines.join(lineEndChar);

    return new vsc.TextEdit(editRange, newText);
  });
}

/**
 * Apply a set of edits to the active text editor. There should be an active
 * text editor and it should not be readonly.
 * @param edits The edits to apply.
 * @returns Promise resolving to a boolean indicating `true` if the edits
 * applied successfully.
 */
export async function applyEditsToCurrentEditor(
  edits: vsc.TextEdit[]
): Promise<boolean> {
  return (
    vsc.window.activeTextEditor?.edit((editBuilder): void => {
      edits.forEach((edit) => editBuilder.replace(edit.range, edit.newText));
    }) ?? false
  );
}

/**
 * Activate the extension. Run automatically by VSCode based on
 * the `activationEvents` property in package.json.
 */
export function activate(): void {
  const command = "docstringFormatter.formatDocstrings";

  const selector: vsc.DocumentSelector = {
    scheme: "file",
    language: "python"
  };

  const formatProvider: vsc.DocumentFormattingEditProvider = {
    provideDocumentFormattingEdits: (document) =>
      formatFile(document.fileName).then(hunksToEdits)
  };

  const actionProvider: vsc.CodeActionProvider = {
    provideCodeActions: (document) => [
      {
        title: "Format Docstrings",
        command,
        tooltip: "Format all docstrings in the current document.",
        arguments: [document.fileName]
      }
    ]
  };

  const actionMetadata: vsc.CodeActionProviderMetadata = {
    providedCodeActionKinds: [vsc.CodeActionKind.Source]
  };

  registrations.push(
    vsc.languages.registerDocumentFormattingEditProvider(
      selector,
      formatProvider
    ),
    vsc.languages.registerCodeActionsProvider(
      selector,
      actionProvider,
      actionMetadata
    ),
    vsc.commands.registerCommand(command, (filePath?: string) =>
      formatFile(filePath)
        .then(hunksToEdits)
        .then(applyEditsToCurrentEditor)
    )
  );
}

/**
 * Unregister all registrations stored in the global `registrations` array.
 */
export function unregisterAll(): void {
  while (registrations.length) {
    registrations.pop()?.dispose();
  }
}

/**
 * Deactivate the extension. Runs automatically upon deactivation or uninstall.
 */
export function deactivate(): void {
  unregisterAll();
}

/**
 * Exception thrown when formatting fails.
 */
export interface FormatException {
  message: string;
}
